wallet-api/src/Ledger/Value/TH.hs

```diff
@@ -59,43 +63,83 @@ similar to 'Ledger.Ada' for their own currencies.
 
 -}
 
+-- | How much of a given currency is in a 'Value'
+valueOf :: Q (TExp (Value -> CurrencySymbol -> Int))
+valueOf = [|| 
+  \(Value xs) cur -> 
+      let 
+          go [] = 0 :: Int
+          go ((c, i):xs') = if $$(P.eq) c cur then i else go xs'
+      in go xs
+   ||]
+
+singleton :: Q (TExp (CurrencySymbol -> Int -> Value))
+singleton = [|| \c i -> Value [(c, i)] ||]
+
+mergeWith :: Q (TExp ((Int -> Int -> Int) -> Value -> Value -> Value))
+mergeWith = [||
+  \f (Value ls) r@(Value rs) -> 
+    let 
+      ls' = $$(P.map) (\(c, i) -> (c, (f i ($$(valueOf) r c)))) ls
+      rs' = $$(P.filter) (\(c, _) -> $$(P.not) ($$(P.any) (\(c', _) -> $$(P.eq) c' c) ls)) rs
+      rs'' = $$(P.map) (\(c, i) -> (c, (f 0 i))) rs'
+    in Value ($$(P.append) ls' rs'')
+  ||]
+
+scale :: Q (TExp (Int -> Value -> Value))
+scale = [|| \i (Value xs) -> Value ($$(P.map) (\(c, i') -> (c, $$(P.multiply) i i')) xs) ||]
+
 -- Num operations
 
 plus :: Q (TExp (Value -> Value -> Value))
-plus = [|| \(Value a) (Value b) -> Value ($$(P.plus) a b)||]
+plus = [|| $$(mergeWith) $$(P.plus) ||]
+
+negate :: Q (TExp (Value -> Value))
+negate = [|| $$(scale) (-1) ||]
 
 minus :: Q (TExp (Value -> Value -> Value))
-minus = [|| \(Value a) (Value b) -> Value ($$(P.minus) a b)||]
+minus = [|| $$(mergeWith) $$(P.minus) ||]
 
 multiply :: Q (TExp (Value -> Value -> Value))
-multiply = [|| \(Value a) (Value b) -> Value ($$(P.multiply) a b)||]
-
-divide :: Q (TExp (Value -> Value -> Value))
-divide = [|| \(Value a) (Value b) -> Value ($$(P.divide) a b)||]
+multiply = [|| $$(mergeWith) $$(P.multiply) ||]
 
 zero :: Q (TExp Value)
-zero = [|| Value 0 ||]
-
-negate :: Q (TExp (Value -> Value))
-negate = [|| \(Value i) -> Value ($$(P.multiply) (-1) i) ||]
+zero = [|| Value [] ||]
 
 isZero :: Q (TExp (Value -> Bool))
-isZero = [|| \(Value i) -> $$(P.eq) i 0 ||]
+isZero = [|| \(Value xs) -> $$(P.all) (\(_, i) -> $$(P.eq) 0 i) xs ||]
 
 geq :: Q (TExp (Value -> Value -> Bool))
-geq = [|| \(Value i) (Value j) -> $$(P.geq) i j ||]
+geq = [|| 
+  \ls rs -> 
+    let Value xs = $$(minus) ls rs
+    in
+      $$(P.all) (\(_, i) -> $$(P.geq) i 0) xs ||]
```

Comment:
Speculating here:
```
type Map k v = [(k, v)]
data These a b = This a | That b | These a b
union :: Map k a -> Map k b -> Map k (These a b))
-- some appropriate definition of geq
```
Not sure that would actually end up nicer, though.

Reply:
we would still need `P.all` though, right? Or is it `minus` that you want to get rid of?

Reply:
Yeah it just seems a bit ad-hoc that we happen to be able to use `a-b` as our indicator for the things we care about. It seems better to use a richer datatype although it doesn't matter in the cases here.

Reply:
Okay, I'll make it a bit nicer